chrome/content/zotero/xpcom/itemTreeManager.js

```diff
@@ -0,0 +1,267 @@
+/*
+	***** BEGIN LICENSE BLOCK *****
+
+	Copyright © 2019 Corporation for Digital Scholarship
+					 Vienna, Virginia, USA
+					 https://digitalscholar.org
+
+	This file is part of Zotero.
+
+	Zotero is free software: you can redistribute it and/or modify
+	it under the terms of the GNU Affero General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	Zotero is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU Affero General Public License for more details.
+
+	You should have received a copy of the GNU Affero General Public License
+	along with Zotero.  If not, see <http://www.gnu.org/licenses/>.
+
+	***** END LICENSE BLOCK *****
+*/
+
+const { COLUMNS: ITEMTREE_COLUMNS } = require("zotero/itemTreeColumns");
+
+/**
+ * @typedef {import("../itemTreeColumns.jsx").ItemTreeColumnOption} ItemTreeColumnOption
+ */
+
+class ItemTreeManager {
+	_observerAdded = false;
+	/** @type {ItemTreeColumnOption[]} */
+	_customColumns = [];
+	/** 
+	 * Cached custom columns' data source, dataKey: dataProvider
+	 * @type {Record<string, (item: Zotero.Item, field: string, unformatted?: boolean) => string>}
+	 */
+	_customDataProvider = {};
+	constructor() {
+	}
+
+	/** 
+	 * Register a custom column. All registered columns must be valid, and must have a unique dataKey.
+	 * @param {ItemTreeColumnOption | ItemTreeColumnOption[]} optionOrOptions
+	 * @returns {boolean} Although it's async, resolving does not promise the item trees are updated.
+	 * @example
+	 * ```js
+	 * Zotero.ItemTreeManager.registerColumns(
+	 * {
+	 *     dataKey: 'rtitle',
+	 *     iconPath: 'chrome://zotero/skin/tick.png',
+	 *     label: 'reversed title',
+	 *     dataProvider: (item, field) => {
+	 *         return item.getField('title').split('').reverse().join('')}
+	 *     }),
+	 * });
+	 * ```
+	 */
+	async registerColumns(optionOrOptions) {
```

Comment:
This and all of the above arguments should just be in plural. So "options", "dataKeys", etc.